Enable access to group name

ajbozarth opened the pull request:
Adds two functions `tls1_group_id2name` which returns the `tlsname` for a given `group_id` and `SSL_get0_group_name` which surfaces that name to the "user" 

This is a follow up to my previously rejected PR #20318 and the resulting discussion. It is related to #20594 but I am unsure if it addresses that issue at its core.

Since I am new to this project I would appreciate direction on if I went about surfacing the `tlsname` the correct way here, how I should document this new functionality, and how/where to test it.

I am aware I will need to submit a ICLA and a CCLA (If IBM hasn't submitted one already) before this can be merged.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

paulidale commented on ssl/t1_lib.c:
```diff
@@ -755,6 +755,14 @@ const TLS_GROUP_INFO *tls1_group_id_lookup(SSL_CTX *ctx, uint16_t group_id)
+const char *tls1_group_id2name(SSL_CTX *ctx, uint16_t group_id)
+{
+    TLS_GROUP_INFO *tls_group_info = tls1_group_id_lookup(ctx, group_id);
+    if (tls_group_info == NULL)
```
Out coding style requires a blank line between declarations and code.

paulidale reviewed (changes requested):
We require documentation for the new function and ideally a test case.

ajbozarth commented:
A few updates:
- I've address the style issues after reviewing https://www.openssl.org/policies/technical/coding-style.html
- I've added a doc file based on reading doc files for similar surrounding functions.
- I will submit my ICLA on Friday (I will have access to a printer/scanner Thursday night)
- I've confirmed that IBM has previously submitted a CCLA
- I am in the process of figuring out where/how to test this. I will be adding/updating a test in `test/sslapitest.c`

hlandau commented on doc/man3/SSL_get_curve_name.pod:
```diff
@@ -0,0 +1,41 @@
+the key agreement of the current TLS session establishment.
+
+Note that the return value is valid only during the lifetime of the
+SSL object I<ssl>.
```
What happens if the handshake hasn't been initiated or completed yet? If this returns NULL in that case, it would be good to mention it.

ajbozarth replied:
Good point, I'm not actually sure if it would return NULL or just break. I will make sure it returns NULL and update the doc to reflect it.

hlandau reviewed (changes requested):
One small nit, other than that looks good.

ajbozarth commented:
I've added some test updates, but am not confident I added enough. Feedback on my tests would be appreciated.

Also as an FYI, I emailed in my ICLA today.

paulidale reviewed (approved):
Tests look good enough.

t8m commented:
CI failures are relevant

ajbozarth commented:
I believe I've addressed the CI failures, there was one windows build failure I was not sure about though.

If you could re-trigger the CI for me (since I'm not whitelisted)

t8m commented:
Please rebase the PR against fresh master branch and drop the merge commit. Merge commits are not allowed.

ajbozarth commented:
> Please rebase the PR against fresh master branch and drop the merge commit. Merge commits are not allowed.

Since most of my commits were just `wip something` I decided to just patch the diff and force push it. 

Everything should be fine now, though I'm unsure why my CCLA situation is still unresolved (my ICLA was accepted over a week ago).

ajbozarth commented:
Closed and reopened to grab my CLA

ajbozarth commented on doc/man3/SSL_get_curve_name.pod:
```diff
@@ -0,0 +1,43 @@
+
+ #include <openssl/ssl.h>
+
+ const char *SSL_get_curve_name(SSL *s);
```
I've pushed an update with the name change. I left the function where it is, but if you'd like me to move the two functions to `ssl_lib.c` I can do that.

ajbozarth replied:
also if you could double check the doc wording, I just replaced the word "curve" with "group" but am unsure if I need to reword the entire sentence instead for the correct meaning

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

t8m commented:
@hlandau @paulidale please reconfirm

ajbozarth commented:
FYI I saw a merge conflict message so I rebased on latest master and force pushed it

Tarnum-tst commented:
'''
SSL_get_curve_name                      ?	3_2_0	EXIST::FUNCTION:
'''
should be removed from util/libssl.num

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

paulidale commented:
Needs the fix up before merge.  Approved anew once that's done.

ajbozarth commented:
just pushed the requested fix

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Squashed commits and merged to master branch. Thank you for your contribution.

ajbozarth commented:
Follow up Question:
Is there any way to know when this will be part of a release? From my research it would assume it needs to wait for 3.2, which doesn't seem to have an estimate yet afaik.
I'm working on some upstream usage of the new function and would prefer to have a timeline

paulidale commented:
No release date for 3.2 has been set at this stage.
I'd expect late this year, but not promises.
